remote_command_execution_vulnerability.py

```diff
@@ -17,13 +17,46 @@
 import shutil
 import tarfile
 import requests
+import sys
+import re
+import time
+import random
+import hashlib
 
-router_ip_address = "192.168.31.1"
-router_ip_address = input("Router IP address [press enter for using the default {}]: ".format(router_ip_address)) or router_ip_address
+router_ip_address="miwifi.com"
+#router_ip_address = "192.168.31.1"
+router_ip_address = input("Router IP address [press enter for using the default '{}']: ".format(router_ip_address)) or router_ip_address
 
 # get stok
-stok = input("stok: ")
-# stok = "eeb59f33a51cd46649cd4ad1e3f50ecf"
+def get_stok(router_ip_address):
+    try: 
+        r0 = requests.get("http://{router_ip_address}/cgi-bin/luci/web".format(router_ip_address=router_ip_address))
+    except:
+        print ("Xiaomi router not found...")
+        sys.exit(1)
+    try:	
+        mac = re.findall(r'deviceId = \'(.*?)\'', r0.text)[0]
+    except:
+        print ("Xiaomi router not found...")
+        sys.exit(1)
+    key = re.findall(r'key: \'(.*)\',', r0.text)[0]
+    nonce = "0_" + mac + "_" + str(int(time.time())) + "_" + str(random.randint(1000, 10000))
+    router_password = input("Enter router admin password: ")
+    account_str = hashlib.sha1((router_password + key).encode('utf-8')).hexdigest()
+    password = hashlib.sha1((nonce + account_str).encode('utf-8')).hexdigest()
+    data = "username=admin&password={password}&logtype=2&nonce={nonce}".format(password=password,nonce=nonce)
+    r1 = requests.post("http://{router_ip_address}/cgi-bin/luci/api/xqsystem/login".format(router_ip_address=router_ip_address), 
+        data = data, 
+        headers={"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:65.0) Gecko/20100101 Firefox/65.0",
+            "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8"})
+    try:
+        stok = re.findall(r'"token":"(.*?)"',r1.text)[0]
+    except:
+        print("Failed to get stok in login response '{}'".format(r1.text))
+        sys.exit(1)
+    return stok
+
+stok = get_stok(router_ip_address)

```

Comment:
Would it be possible to ask the STOK directly to the user if this function fails? As backup